### ISSUE TITLE
New boolean operations issue.

### ISSUE THREAD

**Issue by lehni** [paperjs/paper.js#964]
@iconexperience [this one's again from your test-case](http://sketch.paperjs.org/#S/XVTLbhxVFPyVljcOUs/VeT9isWKNhJQlYTFMOrHlYcYaGyJA/Dt1bQwkGo3Ufe95VNWp039enfa/bFdvr97db0+H26v16nD+MN9/21+Wh/3TLS/fLqft8/IDnt9cf2/cQ0jNPLyUnGPVsqEWlJpNLRZx8FFmyRrMxpLsqw5lhBci3NxVlhpICGkUCo6SQkwQFVE3ugghfekRgfoZjR8jAsmZaVZRbl15oGEyH7yT2a1a1x0NUZkvEqyenS6LDgZqYtYSrlJbdzyaiKO7ndEfBJYYzeIzoAugDFE9SDLYMl25UvugIOMIbzXVJNJYdyGjlbiVmXBDaQv7IKvi0sxi8pB1JwEKSHZjUjSOXNhGNbKaDF1AY0bZFNmj2pTxP3oNbUVxIessA0nWkVApyYTTIZpAC2gV3AVGkYbqsfIIKeOgTq1ozlwE82IDIIgNbStzFYEY5ajfJShbuuwEMiZDBgzFGKPLVSkGdXiHNRC3+2GXI6EwJEojdGBewQidgRMHnlSBYrANUKdM5SFuWKyCo+TkhhusLX3ZacEC6lSeNY/RMQcx+KRjcCh6mLACvAtOK2vAjTXHLDsZZ1BM7MYD7qSG9gTpURzemZNQEowaHaqWXdDwVIUEz3MzX8EGJpm+mSC5wv64/ubm/el1G+SrbXCAyWl6wQgxgpUneYAIeI6mJeOwczidsyJEHGwJYHaYgQCwgLkDNwhCcDhNuKNIBXbsslViwD0hgSnnpJaLCohg2+D4xv4lIwrmA3fHomXD43ZETyijjYLgCtvCe95jioE1AArYPA86bdasmJLohLjuFOuUWFGYLEQTKmECjXhQghOx1bFaDhES42kod6kXTI2EUgZKfBQmJG8uR2uEUL7o+P70/E0Zj0+X8/323fl4vkDR68v24frm5U6+vvt02bbT6+1r5rvD/nh3+oT7/+f8d/pxf3zcnvvNyV22x38iefx6unva3jxnvQDC5fh4dzz+2/D37Xg8f54dv7ga++PD7R4BWAu/wYfy58u2v384352eHq/e/vjTX38D).
```js
var path1 = new Path('M419.20344565830516,384.34607379092466c5.844713611412715,3.313058909545532 8.737629565616828,3.6008009941920136 9.663847696961,3.5457774486854987c0.4249875971154893,-0.23215482613579752 3.134401138218834,-1.9001699951828073 6.912521889816844,-9.027614775318739c33.358079343370036,-62.930193110079074 15.048818377810562,-263.53585410399967 14.89110904614597,-264.2035689431943l58.39331120497843,-13.792070421750992c0.38461989956749676,1.6284160973869177 24.314056080847877,221.98511298279283 -20.271816008415897,306.09659649109955c-7.779727674038611,14.676497403570863 -19.843728834889646,29.87171953249475 -38.60350857895327,37.0189975138764c-20.692089248499826,7.88346749760683 -41.830098370105475,3.183730282850888 -60.57334789311045,-7.440826188961864z');
var path2 = new Path('M450.6709242911992,114.66066752097116c-5.7317866225195075,-24.267422495749997 -2.0221968032280984,26.377626042706837 32.186077599051714,22.94975715794014l-5.982390225081872,-59.70101345199127c34.20913357233178,-3.427954978623788 37.91990661944476,47.22024138695528 32.18962383106816,22.95918587255207z');

path1.strokeColor = 'red';
path2.strokeColor = 'green';
path1.strokeScaling = path2.strokeScaling = false;

var res = path1.unite(path2);

res.fillColor = 'yellow';
res.fillColor.alpha = 0.25;
```

It works online but doesn't with the latest develop branch. It could be that I was a bit too eager removing my edge-case code...

**Comment by lehni** [comment]
The good news is that it's the only failing case.

**Comment by lehni** [comment]
So I've been looking into this error here, and it turns out that the removed edge-case handling code  `isCrossing()`was actually returning a false result here as well, which then in turn has prevented this issue form happening. But the origin of the issue must be elsewhere.

**Comment by iconexperience** [comment]
Here is a simplification of the issue:

![image](https://cloud.githubusercontent.com/assets/2122773/13026964/8343198e-d23c-11e5-9ce0-a9f5dcb48cd4.png)

```
var path1 = new paper.Path({segments:[
    [450.67092429122897, 114.66066752109714, 5, 5, 0, 0],
    [509.0642354962074, 100.86859709934615, 0, 0, 5, 5]], closed:true});
var path2 = new paper.Path({segments:[
    [450.6709242911992, 114.66066752097116, 0, 0, ],
    [482.8570018902509, 137.6104246789113, -34.20827440227981, 3.4278688847666956, 0, 0],
    [476.874611665169, 77.90941122692003, 0, 0, ],
    [509.0642354962372, 100.8685970994721, 0, 0]], closed:true});

path1.strokeColor = 'red';
path2.strokeColor = 'green';
path1.strokeScaling = path2.strokeScaling = false;
var res = path1.unite(path2);

res.fillColor = 'yellow';
res.fillColor.alpha = 0.25;
```

**Comment by iconexperience** [comment]
And here are the found intersections. The red dots are crossings:

![image](https://cloud.githubusercontent.com/assets/2122773/13027002/7368aab8-d23e-11e5-9ae5-1f6208a53481.png)

It's interesting that the sketch finds the same intersections and the crossings are the same, yet it does not fail on this example, while the current develop version fails.

**Comment by lehni** [comment]
I think what finally comes to bite us here is the fact that so far we consider all overlaps as crossings without any further checks. There's already a TODO about this, time to finally implement it, it seems:

https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.js#L312

**Comment by lehni** [comment]
Actually, I think I'm wrong here. it shouldn't be a problem in the boolean code that we potentially split in slightly too many places, as long as we do the right thing after. This is probably a bug in `#isCrossing()`. And that explains why you see different behavior on the different versions.

**Comment by lehni** [comment]
Hmm, I think we see multiple issues here. First of all, isn't it weird in your example that the left intersection is no crossing, and the right one is? The right one really doens't look like one, or am I wrong?

But the actual issue is hidden: The green shape is split into two by `#resolveCrossings()`. The point wher it intersects with itself produces two intersections with the other shape (because there are now two paths). And one of them is considered to be a crossing, while the other one is not. That is what I believe the source of the problem here.

**Comment by iconexperience** [comment]
Why is the red circle in the middle so far off the line? This is strange.

**Comment by lehni** [comment]
Like I said, I moved it so you can see both : )

**Comment by iconexperience** [comment]
Ooops :blush:

**Comment by iconexperience** [comment]
@lehni Sorry can't concentrate today, there is too much going on around me. Hope that I can do a little work tomorrow.

**Comment by iconexperience** [comment]
I am trying againto solve this issue again. As a first step <a href="http://sketch.paperjs.org/#S/nZLbTsMwDIZfpeoNQ6oi202cZhNXvAASl9suyuYdtKyd0gJCaO+OB+yE4IaqldL4s/37T97zpt5KPswfN9LPVnmRz9r54f+lTtmu7leY3WWNvOp6J8k86M7gvZPlVpq+G44nadJk+oytA8MeAlkKSFQFX2SI1jADs3eEEDzaIgN9p8UpzUEwwJZKZwMTeCUQwFRcuaDVQmkZ3QVvKy2ABNNpkc1i28l82Kdn2d+ODshRM/1DM4ZA15IPipGvmpOpnAfAKgCpdOVLbxjBkmVfaZHyEvdsKq8DILNDVtp7EyBYVIM4EED5pxOlpx9OWE/4+9SH7/OkTNendiP3bWyTOnCTZH4zOkbpZ3SZRJpz/Jj9OKvjulkqcZl13l3UsZOT20m6bxLNc7PuZfCZdRamgFmsYzy1fZMY29evvldBU8fdqlYEDLmRXsSnJPVm16710PLheLr/AA==">here is a version with only straight lines that still has the same issue.</a>

**Comment by iconexperience** [comment]
And here is a further simplification. I have split the green shape at the self intersection and removed the right part.

<a href="http://sketch.paperjs.org/#S/lVLLTsMwEPwVK5cWEVlrx4+4FSd+AIlj20NIl7SqsSs7BSHEv7MB0lIQSESJZHtm1jObfSlC84DFrLjdYd9uirJo43rYPzaJ7Zt+I9gVC/jEbmg9fcnYPWDo82yxTMvA6FkoDdxYcFJJJ6SsnS2ZEIobA8ZYLQU4K9SqPAo0OA5GyUorZyRYRXwAXptaO6rjKmWE/sJXNRBDwmo8WZWs9THjetanA75ezAdg9Cv/5Vc4J8/tDm6FKRkc3zMrktfaAojagaQgpK0sNwKUVMbWVLD6TWodRTQVXawt9WmQguXWKFeRFU3t0GfSP9MO3/vf4blPcYfX0cdEyScJ15P5iMrvaJcQwwkf1bdt47ehI8ZX1en0vvEZj11OmD+Zgh/Ctsfpu+pkrI0hR4/cx276Qeuwv04xZyqXP9ncY+j6Dbtky4K1I7gsTmXonm/+w8H7+Qjdb70/BntG7+PT5CfIG7/fNEQBLvWcxvsuYbPbxy0NRTFbrF7fAA==">Here is the result</a>

This does not actually fail, but it finds three crossings between two paths, which cannot be correct.

**Comment by iconexperience** [comment]
And some more details:

- `CurveLocation.insert` merges the locations of the two intersections to one, therefore only one intersection/crossing is found.
- `isCrossing()` returns true, because the `t`values are far enough from the end points (although just slightly), and the curves are not merely touching tangentially (obviously).
So there is an inconsistency between there two.

**Comment by iconexperience** [comment]
I think the problem is in <a href="https://github.com/paperjs/paper.js/blob/develop/src/path/CurveLocation.js#L529">`CurveLocation.js line 529`</a>. We should probably not simply calculate the distance between the points, but check if they are on different curves and if so calculate the distance as the sum of the distances to the segment point between the two points. The same is true for  <a href="https://github.com/paperjs/paper.js/blob/develop/src/path/CurveLocation.js#L328">`CurveLocation.js line 328`</a>

Here is what I mean for our test case. Currently we use the direct distance between the two intersections:
![image](https://cloud.githubusercontent.com/assets/2122773/15176261/87bbf828-176a-11e6-9fdf-ea51dbd5c6b8.png)

But we should be using the distances from the curve point:
![image](https://cloud.githubusercontent.com/assets/2122773/15176302/bf2465fc-176a-11e6-82d0-be5186e796c7.png)

**Comment by iconexperience** [comment]
Ignore my previous post, I was wrong. The distance is already calculated correctly at <a href="https://github.com/paperjs/paper.js/blob/develop/src/path/CurveLocation.js#L346">`CurveLocation.js line 346/347`</a>.

This means the bug is probably in `CurveLocation.isCrossing()`.

**Comment by iconexperience** [comment]
Here is why our algorithm currently believes the intersection is a crossing. For better understanding I have slightly changed the direction of the the lines (in the example above the top red and green lines are almost parallel).

![image](https://cloud.githubusercontent.com/assets/2122773/15180724/a101ee82-1783-11e6-82a0-50689a87963b.png)

- Both intersections are merged into one, because they are very close to each other. This is fine.
- Since the intersection is very close to P1 and P2, the algorithm in `isCrossing` treats the intersection as if P1 and P2 are the same. With this assumption the intersection actually becomes a crossing:

![image](https://cloud.githubusercontent.com/assets/2122773/15180766/d0c7db4a-1783-11e6-96a7-d21bc801aca4.png)

So the question is how to improve our algorithm to detect these cases correctly.

**Comment by iconexperience** [comment]
Here is a better visualization of the weakness in our algorithm.

![image](https://cloud.githubusercontent.com/assets/2122773/15181413/7afae3ca-1787-11e6-9ce9-a6c8a26bfad1.png)

**Comment by iconexperience** [comment]
Here is a proposal for cases where straight curves are involved (this should be the majority of these edge cases). Instead of comparing the tangential angles to determine the order of the curves we could compare the side of the end point. This would need a little refactoring in <a href="https://github.com/paperjs/paper.js/blob/develop/src/path/CurveLocation.js#L444">CurveLocation.js lines 444 to 480</a>, but it could work.

**Comment by lehni** [comment]
@iconexperience this sounds good. But it wouldn't actually solve the example in this issue above, right?

**Comment by iconexperience** [comment]
@lehni Yes, this may be completely irrelevant for this issue and I will hav another look. Acutally I am quite deep into the the boolean code, so I should use this momentum.

Thanks for all the work that you have done today.

**Comment by iconexperience** [comment]
@lehni Can you confirm that this was fixed as a side effect of 719392d ? :wink:

**Comment by iconexperience** [comment]
Here are the calculated windings for the curves. Red means 2, green means 1
![image](https://cloud.githubusercontent.com/assets/2122773/15983656/6283d816-2faf-11e6-9660-0cf9fd06a9eb.png)
I think the red curve where the arrow points at is wrong, it should be green.

**Comment by iconexperience** [comment]
And here is where the winding has been calculated. Again, red means winding 2, green means winding 1.
![image](https://cloud.githubusercontent.com/assets/2122773/15983673/fc042e5a-2faf-11e6-90d2-dd8e8d2af0b3.png)

(there are no hidden red dots under the green ones, I checked).

The red curve marked before probably got it's winding through propagation. Since a chain always runs from crossing to crossing, this belongs to the same chain as the inner curves. Maybe a chain should only run from intersection to intersection, no matter if the intersection is a crossing or not?

**Comment by lehni** [comment]
@iconexperience I can't see any errors in my initial sketch here anymore either! You sure there still are issues? : )

**Comment by lehni** [comment]
@iconexperience regarding the chains, where do you see the code that makes the distinction between crossings and intersections there? I can't see it in `propagateWinding()`, and I think it was always behaving the way you describe it.

**Comment by iconexperience** [comment]
The result looks fine, but there is still an "open path" error message. That's probably the red curves.

Regarding the chains, I was probably wrong. I thought the `segment._intersection` exists only for crossing, but of course it's also there is there is an overlap. 

But in your original example the three red curves build one chain.

![image](https://cloud.githubusercontent.com/assets/2122773/15984732/780ed958-2fd5-11e6-885c-4aa575f04123.png)

Why isn't there an intersection at the top right breaking the chaing?

**Comment by lehni** [comment]
I'm not sure... But I don't get that message. Are you sure you're working with the latest code? Which example are you testing with?

**Comment by iconexperience** [comment]
Oh dear, I had a second test in my code that cause the message (but still a correct result). Sorry.

Anyway, let's just close this.

**Comment by lehni** [comment]
Not so fast... This was apparently fixed by 648beb33e90821843f1b19143aaaf04f119bcc6d, but that's causing a shit-ton of new issues, see here: https://github.com/paperjs/paper.js/issues/1054#issuecomment-225354937

**Comment by lehni** [comment]
Ok, I feel this is safe to close now. I will add a unit test for it and then close. Woop!
